gateways/core/artifact/s3.go

```diff
@@ -0,0 +1,186 @@
+	SecretKey     apiv1.SecretKeySelector `json:"secretKey,omitempty" protobuf:"bytes,4,opt,name=secretKey"`
+}
+
+type S3EventConfig struct {
+	Endpoint string                      `json:"endpoint,omitempty" protobuf:"bytes,1,opt,name=endpoint"`
+	Bucket   string                      `json:"bucket,omitempty" protobuf:"bytes,2,opt,name=bucket"`
+	Region   string                      `json:"region,omitempty" protobuf:"bytes,3,opt,name=region"`
+	Event    minio.NotificationEventType `json:"event,omitempty" protobuf:"bytes,4,opt,name=event"`
+	filter   S3Filter                    `json:"filter,omitempty" protobuf:"bytes,5,opt,name=filter"`
+}
+
+// S3Filter represents filters to apply to bucket nofifications for specifying constraints on objects
+type S3Filter struct {
+	Prefix string `json:"prefix" protobuf:"bytes,1,opt,name=prefix"`
+	Suffix string `json:"suffix" protobuf:"bytes,2,opt,name=suffix"`
+}
+
+// getSecrets retrieves the secret value from the secret in namespace with name and key
+func (s *s3) getSecrets(client *kubernetes.Clientset, namespace string, name, key string) (string, error) {
+	secretsIf := client.CoreV1().Secrets(namespace)
+	var secret *apiv1.Secret
+	var err error
+	_ = wait.ExponentialBackoff(common.DefaultRetry, func() (bool, error) {
+		secret, err = secretsIf.Get(name, metav1.GetOptions{})
+		if err != nil {
+			if !common.IsRetryableKubeAPIError(err) {
+				return false, err
+			}
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return "", err
+	}
+	val, ok := secret.Data[key]
+	if !ok {
+		return "", fmt.Errorf("secret '%s' does not have the key '%s'", name, key)
+	}
+	return string(val), nil
+}
+
+func (s *s3) listenToNotifications(artifact *S3Artifact) {
+	// retrieve access key id and secret access key
+	accessKey, err := s.getSecrets(s.gatewayConfig.Clientset, s.gatewayConfig.Namespace, artifact.AccessKey.Name, artifact.AccessKey.Key)
+	if err != nil {
+		panic(fmt.Errorf("failed to retrieve access key id %s", artifact.AccessKey.Name))
+	}
+	secretKey, err := s.getSecrets(s.gatewayConfig.Clientset, s.gatewayConfig.Namespace, artifact.SecretKey.Name, artifact.SecretKey.Key)
+	if err != nil {
+		panic(fmt.Errorf("failed to retrieve access key id %s", artifact.SecretKey.Name))
+	}
+	s.gatewayConfig.Log.Info().Str("accesskey", accessKey).Str("secretaccess", secretKey).Msg("minio secs")
+
+	minioClient, err := minio.New(artifact.S3EventConfig.Endpoint, accessKey, secretKey, !artifact.Insecure)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create a done channel to control 'ListenBucketNotification' go routine.
+	doneCh := make(chan struct{})
+
+	// Indicate to our routine to exit cleanly upon return.
+	defer close(doneCh)
+
+	// Listen for bucket notifications
+	for notificationInfo := range minioClient.ListenBucketNotification(artifact.S3EventConfig.Bucket, artifact.S3EventConfig.filter.Prefix, artifact.S3EventConfig.filter.Suffix, []string{
```

Comment:
I see that you're using `ListenBucketNotifiation`.. I previously used this and removed it from my code because it is currently only supported by `minio` s3 and is not part of the standard s3 api. Obviously this makes the code a lot cleaner as you no longer need a stream, but shouldn't we be able to support all s3 sources and not just minio?